Trait not implemented for ProfilingStructGenerator

Hello,
The compilation fails on the build script with the following output:

error[E0277]: the trait bound `gl_generator_profiling_struct::ProfilingStructGenerator: gl_generator::generators::Generator` is not satisfied
  --> lib/build.rs:20:10
   |
20 |         .write_bindings(ProfilingStructGenerator, &mut file_gl)
   |          ^^^^^^^^^^^^^^ the trait `gl_generator::generators::Generator` is not implemented for `gl_generator_profiling_struct::ProfilingStructGenerator`
error: aborting due to previous error
For more information about this error, try `rustc --explain E0277`.
error: Could not compile `gl`.

So in short it tells me a trait is not implemented for the generator..

However I can see such implementation in the source:
https://github.com/Nercury/gl_generator_profiling_struct-rs/blob/master/src/lib.rs#L70

The only thing is that it's 
`impl gl_generator::Generator for` not `gl_generator::generators::Generator`
They differ. Could be a typo?

I saw the crate be used in "The openGl tutorial" and wanted to use it myself to have some guidelines.
Did something change? I use the '0.1.2' version where on the tutorial it is '0.1'.
Should I roll back or not use this crate at all because it's just for demonstration?

Hi,

First of all, thank you for report.

I don't know exactly what is going on, and won't have time to investigate
for at least few weeks. But I will look into it once I have time just to
make sure everything keeps compiling.


On Sun, 4 Aug 2019, 04:56 Kalocsai Richárd, <notifications@github.com>
wrote:

> Hello,
> The compilation fails on the build script with the following output:
>
> error[E0277]: the trait bound gl_generator_profiling_struct::ProfilingStructGenerator:
> gl_generator::generators::Generator is not satisfied
> --> lib/build.rs:20:10
> |
> 20 | .write_bindings(ProfilingStructGenerator, &mut file_gl)
> | ^^^^^^^^^^^^^^ the trait gl_generator::generators::Generator is not
> implemented for gl_generator_profiling_struct::ProfilingStructGenerator
> error: aborting due to previous error
> For more information about this error, try rustc --explain E0277.
> error: Could not compile gl.
>
> So in short it tells me a trait is not implemented for the generator..
>
> However I can see such implementation in the source:
>
> https://github.com/Nercury/gl_generator_profiling_struct-rs/blob/master/src/lib.rs#L70
>
> The only thing is that it's
> impl gl_generator::Generator for not gl_generator::generators::Generator
> They differ. Could be a typo?
>
> I saw the crate be used in "The openGl tutorial" and wanted to use it
> myself to have some guidelines.
> Did something change? I use the '0.1.2' version where on the tutorial it
> is '0.1'.
> Should I roll back or not use this crate at all because it's just for
> demonstration?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Nercury/gl_generator_profiling_struct-rs/issues/1?email_source=notifications&email_token=AAAZ5RHC75FWUTENTEDQR3TQCYZMTA5CNFSM4IJEHOEKYY3PNVWWK3TUL52HS4DFUVEXG43VMWVGG33NNVSW45C7NFSM4HDG44XA>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAZ5RA2R5YA6YFHGZZSUQDQCYZMTANCNFSM4IJEHOEA>
> .
>

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -8,7 +8,8 @@ repository = "https://github.com/Nercury/gl_generator_profiling_struct-rs"
 keywords = ["open_gl", "gl", "profiling", "debug"]
 categories = ["api-bindings", "rendering::graphics-api"]
 documentation = "https://docs.rs/gl_generator_profiling_struct"
+edition = "2018"
 
 [dependencies]
-gl_generator = { version = "0.9", features = ["unstable_generator_utils"] }
+gl_generator = { version = "0.14.0", features = ["unstable_generator_utils"] }
 

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -72,14 +72,14 @@ impl gl_generator::Generator for ProfilingStructGenerator {
         where
             W: io::Write,
     {
-        try!(write_helper(dest));
-        try!(write_header(dest));
-        try!(write_type_aliases(registry, dest));
-        try!(write_enums(registry, dest));
-        try!(write_fnptr_struct_def(dest));
-        try!(write_panicking_fns(registry, dest));
-        try!(write_struct(registry, dest));
-        try!(write_impl(registry, dest));
+        write_helper(dest)?;
+        write_header(dest)?;
+        write_type_aliases(registry, dest)?;
+        write_enums(registry, dest)?;
+        write_fnptr_struct_def(dest)?;
+        write_panicking_fns(registry, dest)?;
+        write_struct(registry, dest)?;
+        write_impl(registry, dest)?;
         Ok(())
     }
 }
@@ -197,15 +197,15 @@ fn write_type_aliases<W>(registry: &Registry, dest: &mut W) -> io::Result<()>
     where
         W: io::Write,
 {
-    try!(writeln!(
+    writeln!(
         dest,
         r#"
         pub mod types {{
             #![allow(non_camel_case_types, non_snake_case, dead_code, missing_copy_implementations)]
     "#
-    ));
+    )?;
 
-    try!(generators::gen_types(registry.api, dest));
+    generators::gen_types(registry.api, dest)?;
 
     writeln!(dest, "}}")
 }
@@ -216,7 +216,7 @@ fn write_enums<W>(registry: &Registry, dest: &mut W) -> io::Result<()>
         W: io::Write,
 {
     for enm in &registry.enums {
-        try!(generators::gen_enum_item(enm, "types::", dest));
+        generators::gen_enum_item(enm, "types::", dest)?;
     }
 
     Ok(())
@@ -287,22 +287,22 @@ fn write_struct<W>(registry: &Registry, dest: &mut W) -> io::Result<()>
     where
         W: io::Write,
 {
-    try!(writeln!(
+    writeln!(
         dest,
         "
         #[allow(non_camel_case_types, non_snake_case, dead_code)]
         #[derive(Clone)]
         pub struct {api} {{",
         api = generators::gen_struct_name(registry.api)
-    ));
+    )?;
 
     for cmd in &registry.cmds {
         if let Some(v) = registry.aliases.get(&cmd.proto.ident) {
-            try!(writeln!(dest, "/// Fallbacks: {}", v.join(", ")));
+            writeln!(dest, "/// Fallbacks: {}", v.join(", "))?;
         }
-        try!(writeln!(dest, "pub {name}: FnPtr,", name = cmd.proto.ident));
+        writeln!(dest, "pub {name}: FnPtr,", name = cmd.proto.ident)?;
     }
-    try!(writeln!(dest, "_priv: ()"));
+    writeln!(dest, "_priv: ()")?;
 
     writeln!(dest, "}}")
 }
@@ -312,7 +312,7 @@ fn write_impl<W>(registry: &Registry, dest: &mut W) -> io::Result<()>
     where
         W: io::Write,
 {
-    try!(writeln!(dest,
+    writeln!(dest,
                   "impl {api} {{
             /// Load each OpenGL symbol using a custom load function. This allows for the
             /// use of functions like `glfwGetProcAddress` or `SDL_GL_GetProcAddress`.
@@ -323,7 +323,7 @@ fn write_impl<W>(registry: &Registry, dest: &mut W) -> io::Result<()>
             #[allow(dead_code, unused_variables)]
             pub fn load_with<F>(mut loadfn: F) -> {api} where F: FnMut(&'static str) -> *const __gl_imports::raw::c_void {{
                 #[inline(never)]
-                fn do_metaloadfn(loadfn: &mut FnMut(&'static str) -> *const __gl_imports::raw::c_void,
+                fn do_metaloadfn(loadfn: &mut dyn FnMut(&'static str) -> *const __gl_imports::raw::c_void,
                                  symbol: &'static str,
                                  symbols: &[&'static str])
                                  -> *const __gl_imports::raw::c_void {{
@@ -340,10 +340,10 @@ fn write_impl<W>(registry: &Registry, dest: &mut W) -> io::Result<()>
                     do_metaloadfn(&mut loadfn, symbol, symbols)
                 }};
                 {api} {{",
-                  api = generators::gen_struct_name(registry.api)));
+                  api = generators::gen_struct_name(registry.api))?;
 
     for cmd in &registry.cmds {
-        try!(writeln!(
+        writeln!(
             dest,
             "{name}: FnPtr::new(metaloadfn(\"{symbol}\", &[{fallbacks}])),",
             name = cmd.proto.ident,
@@ -355,15 +355,15 @@ fn write_impl<W>(registry: &Registry, dest: &mut W) -> io::Result<()>
                     .join(", "),
                 None => format!(""),
             },
-        ))
+        )?
     }
-    try!(writeln!(dest, "_priv: ()"));
+    writeln!(dest, "_priv: ()")?;
 
-    try!(writeln!(
+    writeln!(
         dest,
         "}}
         }}"
-    ));
+    )?;
 
     for cmd in &registry.cmds {
         let idents = generators::gen_parameters(cmd, true, false);
@@ -387,7 +387,7 @@ fn write_impl<W>(registry: &Registry, dest: &mut W) -> io::Result<()>
                 .concat()
         );
 
-        try!(writeln!(dest,
+        writeln!(dest,
                       "#[allow(non_snake_case, unused_variables, dead_code)]
             #[inline] pub unsafe fn {name}(&self, {params}) -> {return_suffix} {{ \
                 let r = __gl_imports::mem::transmute::<_, extern \"system\" fn({typed_params}) -> {return_suffix}>\
@@ -410,7 +410,7 @@ fn write_impl<W>(registry: &Registry, dest: &mut W) -> io::Result<()>
                     (self.GetError.f)() {{ 0 => inc_call(), r => {{ inc_err(); {println} println!("[OpenGL] ^ GL error triggered: {{}}, {{}}", r, gl_error_to_str(r))}} }}"#, println = println)
                       } else {
                           format!("")
-                      }))
+                      })?
     }
 
     writeln!(

